build/build_kubebuilder.sh

```diff
@@ -13,46 +14,64 @@
-# but we actually want it to under the correct expected package in the GOPATH (/go)
-# - Create the directory to host the code that matches the expected GOPATH package locations
-# - Use /go as the default GOPATH because this is what the image uses
-# - Link our current directory (containing the source code) to the package location in the GOPATH
-
-OWNER="sigs.k8s.io"
-REPO="kubebuilder"
-
-GO_PKG_OWNER=${GOPATH/:*}/src/$OWNER
-GO_PKG_PATH=$GO_PKG_OWNER/$REPO
+# This script runs goreleaser using the build/.goreleaser.yml config.
+# While it can be run locally, it is intended to be run by cloudbuild
+# in the goreleaser/goreleaser image.
 
-mkdir -p $GO_PKG_OWNER
-ln -sf $(pwd) $GO_PKG_PATH
+# echo_run echos then evaluates all args.
+function echo_run() {
+  echo $@
+  # Ensure process substitution is evaluated with eval.
+  eval $@
+}
 
-# When invoked in container builder, this script runs under /workspace which is
-# not under $GOPATH, so we need to `cd` to repo under GOPATH for it to build
-cd $GO_PKG_PATH
+# install_notes installs kubebuilder's release notes generator globally with name "notes".
+function install_notes() {
+  local tmp=$(mktemp -d)
+  pushd "$tmp"
+  go mod init tmp
+  go get sigs.k8s.io/kubebuilder-release-tools/notes
+  popd
+  rm -rf "$tmp"
+}
 
+set -o errexit
+set -o pipefail
 
-# NOTE: if snapshot is enabled, release is not published to GitHub and the build
-# is available under workspace/dist directory.
-SNAPSHOT=""
+# SNAPSHOT is set by the CLI flag parser if --snapshot is a passed flag.
+# If not set, release notes are not generated.
+SNAPSHOT=
+# GORELEASER_FLAGS sets up goreleaser flags such that it can be run
+# in local/snapshot/prod mode from the same script.
+# NOTE: if --snapshot is set, release is not published to GitHub
+# and the build is available under $PWD/dist.
+GORELEASER_FLAGS=
 
-# parse commandline args copied from the link below
-# https://stackoverflow.com/questions/192249/how-do-i-parse-command-line-arguments-in-bash?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa
-
-while [[ $# -gt 0 ]]
-do
-key="$1"
-
-case $key in
+while [[ $# -gt 0 ]]; do
+  key="$1"
+  case $key in
     --snapshot)
-    SNAPSHOT="--snapshot"
-    shift # past argument
+    # TODO(estroz): figure out how to generate snapshot release notes with the kubebuilder generator.
+    echo "Running in snapshot mode. Release notes will not be generated from commits."
+    notes="Mock Release Notes for $(git describe --tags --always --broken)"
+    GORELEASER_FLAGS="${key} --release-notes <(echo \"${notes}\")"
+    SNAPSHOT=1
+    shift
     ;;
-esac
+    *)
+    GORELEASER_FLAGS="$GORELEASER_FLAGS ${key}"
+    shift
+    ;;
+  esac
 done
 
-/goreleaser release --config=build/.goreleaser.yml --rm-dist --skip-validate ${SNAPSHOT}
+# Generate real release notes.
+if [ -z "$SNAPSHOT" ]; then
+  tmp_notes="$(mktemp)"
+  trap "rm -f ${tmp_notes}" EXIT
+  install_notes
+  notes | tee "$tmp_notes"
```

Comment:
FYI, there's special flags for betas and whatnot, might want to feed them through here depending on the version being tagged

Reply:
Yeah I was wondering about that. Since this script is run automatically by cloudbuild, how would the flags being passed be determined?

Reply:
You can now set these flags in the `NOTES_FLAGS` env var.

Reply:
ack.  The idea is that we'd determine the flags automatically from the tag this is being released against, but we should really reverse that and have a release notes tooling that that says "given a version, determine if we should be in alpha/beta/whatever mode", but we can do that in a follow-up